packages/fetchai/contracts/oracle/contract.py

```diff
@@ -0,0 +1,130 @@
+
+        :param ledger_api: the ledger API
+        :param contract_address: the contract address
+        :param oracle_address: the address of the oracle
+        :param gas: the gas limit
+        :return: the transaction object
+        """
+        if ledger_api.identifier == EthereumApi.identifier:
+            nonce = ledger_api.api.eth.getTransactionCount(oracle_address)
+            instance = cls.get_instance(ledger_api, contract_address)
+            tx = instance.functions.grantRole(
+                CONTRACT_ROLE, oracle_address
+            ).buildTransaction(
+                {
+                    "gas": gas,
+                    "gasPrice": ledger_api.api.toWei("50", "gwei"),
+                    "nonce": nonce,
+                }
+            )
+            tx = cls._try_estimate_gas(ledger_api, tx)
+            return tx
+        raise NotImplementedError
+
+    @classmethod
+    def get_update_transaction(
+        cls,
+        ledger_api: LedgerApi,
+        contract_address: Address,
+        oracle_address: Address,
+        update_function: str,
+        update_args: Dict[str, Any],
+        gas: int = 50000,
+    ) -> None:
+        """
+        Update oracle value in contract
+
+        :param ledger_api: the ledger apis.
+        :param contract_address: the contract address.
+        :param oracle_address: the oracle address.
+        :param update_function: the oracle value update function.
+        :param update_args: the arguments to the contract's update function.
+        :return: None
+        """
+        if ledger_api.identifier == EthereumApi.identifier:
+            nonce = ledger_api.api.eth.getTransactionCount(oracle_address)
+            instance = cls.get_instance(ledger_api, contract_address)
+            function = getattr(instance.functions, update_function)
+            intermediate = function(*update_args)
+            tx = intermediate.buildTransaction(
+                {
+                    "gas": gas,
+                    "gasPrice": ledger_api.api.toWei("50", "gwei"),
+                    "nonce": nonce,
+                }
+            )
+            tx = cls._try_estimate_gas(ledger_api, tx)
+            return tx
+        raise NotImplementedError
+
+    @staticmethod
+    def _try_estimate_gas(ledger_api: LedgerApi, tx: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Attempts to update the transaction with a gas estimate.
+
+        :param ledger_api: the ledger API
+        :param tx: the transaction
+        :return: the transaction (potentially updated)
+        """
+        try:
+            # try estimate the gas and update the transaction dict
+            gas_estimate = ledger_api.api.eth.estimateGas(transaction=tx)
+            tx["gas"] = gas_estimate
+        except Exception as e:  # pylint: disable=broad-except
+            _default_logger.debug(
+                "[OracleContract]: Error when trying to estimate gas: {}".format(e)
+            )
+
+        return tx
```

Comment:
@DavidMinarsch this is a code similar to other contracts, I think it's an opportunity for refactoring

Reply:
yes